src/cppsim/circuit_optimizer.hpp

```diff
@@ -34,16 +34,31 @@ class DllExport QuantumCircuitOptimizer {
     ////////////////////////////////////////////////////////////
     // for swap insertion
     ////////////////////////////////////////////////////////////
+    struct GateReplacer {
+        std::map<QuantumGateBase*, QuantumGateBase*> replace;
+
+        QuantumGateBase* get_replaced_gate(QuantumGateBase* gate) {
+            if (!replace.count(gate)) return gate;
+            return replace[gate] = get_replaced_gate(replace[gate]);
+        }
+
+        void set_replaced_gate(QuantumGateBase* from, QuantumGateBase* to) {
+            replace[from] = to;
```

Comment:
Overwriting existing references to other references is forbidden in the union-find algorithm.
```cpp
// Sample
QuantumGateBase* A,B,C,D;
set_replaced_gate(A,B);
set_replaced_gate(B,C);
get_replaced_gate(A); // replace[A] is set to C
set_replaced_gate(B,D);
get_replaced_gate(A): // return value expects D, but was C.
```
Please add error handling codes or use normal traversal method without route compression.

```suggestion
            if(replace.count(gate))
              // throw errors here.
            replace[from] = to;
```

Reply:
handled on https://github.com/qulacs/qulacs/pull/661/commits/ca320c54455f3b4bcaad62ece930dffe26c2dbda